### ISSUE TITLE
Using a rule for a specific property when building a class

### ISSUE THREAD

**Issue by crenan** [AutoFixture/AutoFixture#1171]
Hi.

Is it possible to make AutoFixture randomly pick one element in a predefined list (or an array) for a specific property?
Something like:

```C#
public class SomeClass
{
    public char SomeProperty { get; set; }
    ...
}

fixture.Build<SomeClass>()
    .With(cls => cls.SomeProperty, [SomeKindOfARandomlyPicker](new char[]{ 'S', 'N' }))
    .Create();
```

Thanks.

**Comment by Kralizek** [comment]
Hey! Check this

https://github.com/Nybus-project/Nybus/blob/9f2c7d2268787a5a9c7a8a232c303f9d50637871/tests/Tests.Nybus.Engine.RabbitMq/CustomAutoMoqDataAttribute.cs#L35

In this line I set a configuration so that I get a random item from the collection.

**Comment by crenan** [comment]
> Hey! Check this
> 
> https://github.com/Nybus-project/Nybus/blob/9f2c7d2268787a5a9c7a8a232c303f9d50637871/tests/Tests.Nybus.Engine.RabbitMq/CustomAutoMoqDataAttribute.cs#L35
> 
> In this line I set a configuration so that I get a random item from the collection.

Thanks for the reply, @Kralizek.

Correct me if I'm wrong, but the `Fixture.Customizations.Add()` works for any T type been customized, not for a specific property, right?

In my example, if I make a customization over **char**, every **char** property in my classes will have the values defined using the custom generator.
What I really want is for a specific property (either using `.Build<SomeClass>()` or `Customize<SomeClass>()`)

**Comment by Kralizek** [comment]
You can use the same construct when customizing a specific property using With.

**Comment by crenan** [comment]
@Kralizek
Sorry but I can't see a way to create an ISpecimenBuilder as a parameter of the `.With()` method.

If I do this:
```C#
public class SomeClass
{
  public char SomeProperty1 { get; set; }
  public char SomeProperty2 { get; set; }
}

fixture.Customizations.Add(new ElementsBuilder<char>('S', 'N'));
fixture.Create<SomeClass>();
```
Both `SomeProperty1` and `SomeProperty2` will have 'S' or 'N'.

**Comment by Kralizek** [comment]
@crenan 
Sorry, you are right. I could only find a workaround to get to work.

```
char[] validChars = new char[]{'Y', 'N'};

Fixture fixture = new Fixture();

var test = fixture.Build<TestClass>()
			.With(p => p.Property1, (Generator<char> generator) => generator.Where(c => validChars.Contains(c)).First())
			.Create();
```

It's not clean but it works.

**Comment by Kralizek** [comment]
Adding on for future discussion (@zvirja hint hint ;) )

My hope was to use the `Generator<char>` together with the `ElementsBuilder<char>` but it doesn't work.

The following snippet throws at execution time with an exception.
```
// does't work
char[] validChars = new char[] { 'Y', 'N' };

var builder = new ElementsBuilder<char>(validChars);

var generator = new Generator<char>(builder);

Fixture fixture = new Fixture();

var test = fixture.Build<TestClass>()
			.With(p => p.Property1, () => generator.First())
			.Create();
```

The exception: `Unable to cast object of type 'AutoFixture.Kernel.NoSpecimen' to type 'System.Char'`

Stacktrace:
```

at AutoFixture.SpecimenFactory.Create[T](ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\SpecimenFactory.cs:line 23   at AutoFixture.Generator`1.GetEnumerator()+MoveNext() in C:\projects\autofixture\Src\AutoFixture\Generator.cs:line 47   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)   at UserQuery.<>c__DisplayClass5_0.<Main2>b__1(), line 25   at AutoFixture.Dsl.NodeComposer`1.<>c__DisplayClass17_0`1.<With>b__0(ISpecimenContext _) in C:\projects\autofixture\Src\AutoFixture\Dsl\NodeComposer.cs:line 181   at AutoFixture.Kernel.BindingCommand`2.Execute(Object specimen, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\BindingCommand.cs:line 169   at AutoFixture.Kernel.Postprocessor`1.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\Postprocessor.cs:line 285   at AutoFixture.Kernel.CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\CompositeSpecimenBuilder.cs:line 52   at AutoFixture.Kernel.CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\CompositeSpecimenBuilder.cs:line 52   at AutoFixture.Kernel.RecursionGuard.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\RecursionGuard.cs:line 176   at AutoFixture.Kernel.CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\CompositeSpecimenBuilder.cs:line 52   at AutoFixture.Kernel.CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\CompositeSpecimenBuilder.cs:line 52   at AutoFixture.Kernel.RecursionGuard.Create(Object request, ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\Kernel\RecursionGuard.cs:line 176   at AutoFixture.SpecimenFactory.Create[T](ISpecimenContext context) in C:\projects\autofixture\Src\AutoFixture\SpecimenFactory.cs:line 23   at UserQuery.Main2(), line 24   at UserQuery.Main(), line 11
--


```

**Comment by crenan** [comment]
> .With(p => p.Property1, (Generator<char> generator) => generator.Where(c => validChars.Contains(c)).First())

@Kralizek Thank you. It's ugly but works :D .

As for the hint for @zvirja, the first thing I tried was using the `ElementsBuilder<char>('N','S')`, but I was getting the exception you mentioned, so I went to check the code and saw that it's not possible to use an ISpecimenBuilder.

It would be a good thing the possibility to use an ISpecimenBuilder for a specific property, like
```C#
.With(p => p.MyProperty, new SpecimenBuilder(....))
```
so AF could use it to generate values in the post-processing.
